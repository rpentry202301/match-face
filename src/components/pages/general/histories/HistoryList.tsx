"use client";

import GrayButton from "@/components/ui/button/GrayButton";
import Link from "next/link";
import { memo, useEffect, useState } from "react";

interface HistoryListProps {
  click?: boolean;
  month?: string;
  skill?: number[] | undefined;
}
interface TypeProject {
  administratorId: number;
  answered: boolean;
  comment: { context: string };
  createdAt: string;
  createdUser: string;
  deadline: string;
  id: number;
  project: { id: number; name: string; detail: string };
  questionList: [];
  requestAt: string;
  updateAt: string;
  updateUser: string;
  user: {};
}

const HistoryList: React.FC<HistoryListProps> = memo(
  ({ click, month, skill }) => {
    const [selectProject, setSelectProject] = useState<TypeProject[]>([]);

    useEffect(() => {
      async function getSelectedData() {
        const userId = 1;
        const response = await fetch("http://localhost:3000/api/histories", {
          cache: "no-store",
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            user_id: userId,
            month: month,
            skill: skill,
          }),
        });
        if (!response.ok) throw new Error("Failed to fetch data");
        const FetchData = await response.json();
        const selectProject = FetchData.answerRequestList;
        setSelectProject(selectProject);
        // console.log("取得したデータ", selectProject, month, skill);
      }
      getSelectedData();
    }, [click]);

    // 詳細の文字数制限
    const truncateString = (str: any, num: any) => {
      return str.length <= num ? str : str.slice(0, num) + "...";
    };
    // 回答日の表示
    const formedDeadline = (inputData: string): string => {
      const date = new Date(inputData);
      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      return `${year}年${month < 10 ? "0" : ""}${month}月${
        day < 10 ? "0" : ""
      }${day}日`;
    };

    return (
      <div>
        <table className="table-auto border border-collapse my-20 w-[80vw]">
          <thead>
            <tr>
              <th className="border p-4 bg-neutral-300">コメント</th>
              <th className="border  bg-neutral-300">回答日</th>
              <th className="border  bg-neutral-300">案件名</th>
              <th className="border  bg-neutral-300">案件概要</th>
              <th className="border  bg-neutral-300"></th>
            </tr>
          </thead>
          <tbody>
            {selectProject.map((project) => (
              <tr key={project.id}>
                <td className="border text-center p-3">
                  {project.comment.context.length > 0 ? (
                    <span className=" bg-orange text-white p-2 w-40 h-auto w-15 rounded py-1 px-2 text-xs">
                      新着
                    </span>
                  ) : (
                    ""
                  )}
                </td>
                <td className="border text-center p-3">
                  {" "}
                  {formedDeadline(project.updateAt)}
                </td>
                <td className="border text-center px-4">
                  {project.project.name}
                </td>

                <td
                  className="border text-center px-4 "
                  data-test={`projectDetail${project.id}`}
                >
                  {truncateString(project.project.detail, 30)}
                </td>
                <td className="border text-center px-4">
                  <Link href={`result/${project.project.id}`}>
                    <GrayButton
                      label={"詳細"}
                      className="w-15 rounded py-1 px-2 text-xs"
                      value={project.project.id}
                    />
                  </Link>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
);

HistoryList.displayName = "HistoryList";
export default HistoryList;
