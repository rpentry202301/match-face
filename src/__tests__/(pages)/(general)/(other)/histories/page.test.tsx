import { render, screen, cleanup, act, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { mockHistoriesData, skillData } from "./mock";
import HistoriesPage from "@/app/(pages)/(general)/(other)/histories/page";
import HistoryList from "@/components/pages/general/histories/HistoryList";
import userEvent from "@testing-library/user-event";

// スナップショットテスト
describe("スナップショットテスト", () => {
  global.fetch = jest.fn().mockImplementation((url, config) => {
    if (url === "http://localhost:3000/api/histories/skills") {
      return {
        ok: true,
        json: async () => skillData,
      };
    }
    if (url === "http://localhost:3000/api/histories") {
      return {
        ok: true,
        json: async () => mockHistoriesData,
      };
    }
  });
  it("スナップショット", async () => {
    const { container } = render(<HistoriesPage />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});

// 機能・インタラクションテスト
describe("一般ユーザー回答履歴一覧画面", () => {
  global.fetch = jest.fn().mockImplementation((url, config) => {
    if (url === "http://localhost:3000/api/histories/skills") {
      return {
        ok: true,
        json: async () => skillData,
      };
    }
    if (url === "http://localhost:3000/api/histories") {
      return {
        ok: true,
        json: async () => mockHistoriesData,
      };
    }
  });
  beforeEach(async () => {
    await waitFor(() => {
      render(<HistoriesPage />);
    });
  });
  describe("Histories Page", () => {
    it("絞り込み<button>が存在する", async () => {
      await waitFor(() => {
        const element = screen.getAllByRole("button", { name: "絞り込み" });
        expect(element).toBeTruthy();
      });
    });

    it("何も選択せず絞り込みボタンをクリックすると、ユーザーの全件を表示する", async () => {
      await waitFor(async () => {
        const element = screen.getByTestId("submitButton");
        await userEvent.click(element);
        screen.debug();
      });
    });
    it("JavaScriptにチェックがついていないことを確認", async () => {
      let selectedSkill: HTMLElement;
      await waitFor(() => {
        selectedSkill = screen.getByRole("checkbox", { name: "JavaScript" });
        expect(selectedSkill).not.toBeChecked();
      });
    });
    it("JavaScriptにチェックがついていることを確認", async () => {
      let selectedSkill: HTMLElement;
      await waitFor(async () => {
        selectedSkill = screen.getByRole("checkbox", { name: "JavaScript" });
        await userEvent.click(selectedSkill);
        expect(selectedSkill).toBeChecked();
      });
    });
  });
});
