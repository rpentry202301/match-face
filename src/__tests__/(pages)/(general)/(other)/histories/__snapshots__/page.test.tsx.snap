// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`スナップショットテスト スナップショット 1`] = `
<div>
  <div
    class="flex flex-wrap justify-around"
  >
    <section
      class="w-[75vw] border-2 text-center my-[5vh] py-[5vh]"
    >
      <div>
        <div
          class="mb-[2vh]"
        >
          <label
            for="month"
          >
            回答月：
          </label>
          <select
            class="border-2"
            id="month"
            name="month"
          >
            <option
              value=""
            >
              --
            </option>
            <option
              value="2023-01-01"
            >
              2023-01-01
            </option>
            <option
              value="2023-02-01"
            >
              2023-02-01
            </option>
            <option
              value="2023-03-01"
            >
              2023-03-01
            </option>
            <option
              value="2023-04-01"
            >
              2023-04-01
            </option>
            <option
              value="2023-05-01"
            >
              2023-05-01
            </option>
            <option
              value="2023-06-01"
            >
              2023-06-01
            </option>
            <option
              value="2023-07-01"
            >
              2023-07-01
            </option>
            <option
              value="2023-08-01"
            >
              2023-08-01
            </option>
            <option
              value="2023-09-01"
            >
              2023-09-01
            </option>
            <option
              value="2023-10-01"
            >
              2023-10-01
            </option>
            <option
              value="2023-11-01"
            >
              2023-11-01
            </option>
            <option
              value="2023-12-01"
            >
              2023-12-01
            </option>
          </select>
        </div>
        <div
          class="lg:flex justify-center mb-[2vh] max-w-[55vw] ml-[10vw]"
        >
          <legend
            class="lg:w-[215px] lg:h-10 lg:leading-10"
          >
            使用技術：
          </legend>
          <fieldset
            id="skill"
            name="skill"
          />
        </div>
        <div>
          <button
            class="bg-orange drop-shadow-md hover:saturate-150 active:drop-shadow-none active:shadow-inner hover:bg-depp-orange active:bg-depp-orange active:translate-y-0.5 text-white rounded-xl p-2 w-40 h-auto text-lg"
            data-testid="submitButton"
            name="絞り込み"
            type="submit"
          >
            絞り込み
          </button>
        </div>
      </div>
    </section>
  </div>
  <div
    class="flex flex-col items-center h-screen "
  >
    <div>
      <table
        class="table-auto border border-collapse my-20 w-[80vw]"
      >
        <thead>
          <tr>
            <th
              class="border p-4 bg-neutral-300"
            >
              コメント
            </th>
            <th
              class="border  bg-neutral-300"
            >
              回答日
            </th>
            <th
              class="border  bg-neutral-300"
            >
              案件名
            </th>
            <th
              class="border  bg-neutral-300"
            >
              案件概要
            </th>
            <th
              class="border  bg-neutral-300"
            />
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
